File: checkout-core/src/main/java/com/adyen/checkout/core/model/GooglePayDetails.java
package com.adyen.checkout.core.model;

import android.os.Parcel;
import android.support.annotation.NonNull;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * Copyright (c) 2018 Adyen B.V.
 * <p>
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 * <p>
 * Created by timon on 24/04/2018.
 */
public final class GooglePayDetails extends PaymentMethodDetails {
    public static final Creator<GooglePayDetails> CREATOR = new Creator<GooglePayDetails>() {
        @Override
        public GooglePayDetails createFromParcel(Parcel source) {
            return new GooglePayDetails(source);
        }

        @Override
        public GooglePayDetails[] newArray(int size) {
            return new GooglePayDetails[size];
        }
    };

    public static final String KEY_ADDITIONAL_DATA_TOKEN = "additionalData.paywithgoogle.token";

    private String mToken;

    private GooglePayDetails() {
        // Empty constructor for Builder.
    }

    private GooglePayDetails(@NonNull Parcel in) {
        super(in);

        mToken = in.readString();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(mToken);
    }

    @NonNull
    @Override
    public JSONObject serialize() throws JSONException {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put(KEY_ADDITIONAL_DATA_TOKEN, mToken);

        return jsonObject;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        GooglePayDetails that = (GooglePayDetails) o;

        return mToken != null ? mToken.equals(that.mToken) : that.mToken == null;
    }

    @Override
    public int hashCode() {
        return mToken != null ? mToken.hashCode() : 0;
    }

    public static final class Builder {
        private final GooglePayDetails mGooglePayDetails;

        public Builder(@NonNull String token) {
            mGooglePayDetails = new GooglePayDetails();
            mGooglePayDetails.mToken = token;
        }

        @NonNull
        public GooglePayDetails build() {
            return mGooglePayDetails;
        }
    }
}

File: core/src/main/java/com/adyen/checkout/core/components/paymentmethod/GooglePayDetails.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 4/7/2019.
 */
package com.adyen.checkout.core.components.paymentmethod

import com.adyen.checkout.core.common.internal.model.getStringOrNull
import kotlinx.parcelize.Parcelize
import org.json.JSONObject

@Parcelize
data class GooglePayDetails(
    override val type: String?,
    override val sdkData: String? = null,
    val googlePayToken: String? = null,
    val googlePayCardNetwork: String? = null,
    val threeDS2SdkVersion: String? = null,
) : PaymentMethodDetails() {

    companion object {
        private const val GOOGLE_PAY_TOKEN = "googlePayToken"
        private const val GOOGLE_PAY_CARD_NETWORK = "googlePayCardNetwork"
        private const val THREEDS2_SDK_VERSION = "threeDS2SdkVersion"

        @JvmField
        val SERIALIZER: Serializer<GooglePayDetails> = object : Serializer<GooglePayDetails> {
            @Suppress("TooGenericExceptionThrown")
            override fun serialize(modelObject: GooglePayDetails): JSONObject {
                return JSONObject().apply {
                    putOpt(TYPE, modelObject.type)
                    putOpt(SDK_DATA, modelObject.sdkData)
                    putOpt(GOOGLE_PAY_TOKEN, modelObject.googlePayToken)
                    putOpt(GOOGLE_PAY_CARD_NETWORK, modelObject.googlePayCardNetwork)
                    putOpt(THREEDS2_SDK_VERSION, modelObject.threeDS2SdkVersion)
                }
            }

            override fun deserialize(jsonObject: JSONObject): GooglePayDetails {
                return GooglePayDetails(
                    type = jsonObject.getStringOrNull(TYPE),
                    sdkData = jsonObject.getStringOrNull(SDK_DATA),
                    googlePayToken = jsonObject.getStringOrNull(GOOGLE_PAY_TOKEN),
                    googlePayCardNetwork = jsonObject.getStringOrNull(GOOGLE_PAY_CARD_NETWORK),
                    threeDS2SdkVersion = jsonObject.getStringOrNull(THREEDS2_SDK_VERSION),
                )
            }
        }
    }
}
